### ISSUE TITLE
onModuleInit order with import dependencies

### ISSUE THREAD

**Issue by j** [nestjs/nest#7582]
I expect the order of module's init method to be called in this order: `AModule -> BModule -> AppModule`, but I'm getting `AModule -> AppModule -> BModule`

I created a reproduction and tested v7 with the exact code and v7 works as expected.

v7: https://github.com/j/nestjs-issue-7582/tree/v7 (works) `[ 'AModule', 'BModule', 'AppModule' ]`
v8: https://github.com/j/nestjs-issue-7582/tree/main (broken) `[ 'AModule', 'AppModule', 'BModule' ]`

And the code for reference...

```typescript
import { NestFactory } from '@nestjs/core';
import { Module, OnModuleInit } from "@nestjs/common";
import { FastifyAdapter, NestFastifyApplication } from "@nestjs/platform-fastify";

const order: string[] = [];

async function delay(name: string): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(() => {
      console.log(`${name}`);

      order.push(name);

      resolve();
    }, 1000);
  });
}

@Module({
  imports: [],
})
export class AModule implements OnModuleInit {
  async onModuleInit(): Promise<any> {
    await delay('AModule');
  }
}

@Module({
  imports: [AModule],
})
export class BModule implements OnModuleInit {
  async onModuleInit(): Promise<any> {
    await delay('BModule');
  }
}

@Module({
  imports: [BModule],
})
export class AppModule implements OnModuleInit {
  async onModuleInit(): Promise<any> {
    await delay('AppModule');
  }
}

async function bootstrap() {
  const app = await NestFactory.create<NestFastifyApplication>(
    AppModule,
    new FastifyAdapter(),
  );
  await app.listen(6000);

  console.log(order); // ['AModule', 'AppModule', 'BModule'] --- this should be ['AModule', 'BModule', 'AppModule']
}
bootstrap();
```

**Comment by micalevisk** [comment]
https://docs.nestjs.com/fundamentals/lifecycle-events#lifecycle-sequence
![image](https://user-images.githubusercontent.com/13461315/125856053-f6bfed03-f8e7-4a13-b309-9a4ac966f340.png)
they won't change this.

**Comment by j** [comment]
I read that as “called after dependents are called”

**Comment by micalevisk** [comment]
> I read that as “called after dependents are called”

and this is indeed what is happened when I tested here (with v7)

**Comment by j** [comment]
@micalevisk https://github.com/j/nestjs-issue-7582/blob/main/src/main.ts

Reproduced this and order is definitely incorrect.

**Comment by j** [comment]
@micalevisk I can confirm that this works as expected in v7 and incorrect in v8.

v7: https://github.com/j/nestjs-issue-7582/tree/v7 (works)
v8: https://github.com/j/nestjs-issue-7582/tree/main (broken)

**Comment by j** [comment]
Looking at module discovery in v8, the logic definitely has changed so this should be a confirmed bug.  Hopefully it can be fixed soon as some of the libraries I'm using already upgraded to v8 with latest features :-\

**Comment by jmcdo29** [comment]
Thank you for all the research and reproductions for this! It's definitely going to help in building up a test suite around this functionality.

**Comment by micalevisk** [comment]
I've made this simple change to cover your scenario @j on add7eefeecd5597c5dc038daa70c3d5daed932cb (v8.0.3) and the tests are still passing, I didn't understand why.

```diff
@@ -1,3 +1,4 @@
+import { promisify } from 'util';
 import { Injectable, Module, OnModuleInit } from '@nestjs/common';
 import { Test } from '@nestjs/testing';
 import { expect } from 'chai';
@@ -39,6 +40,8 @@ describe('OnModuleInit', () => {
   });
 
   it('should sort modules by distance (topological sort) - DESC order', async () => {
+    const sleep = promisify(setTimeout);
+    const order = [];
     @Injectable()
     class BB implements OnModuleInit {
       public field: string;
@@ -51,7 +54,12 @@ describe('OnModuleInit', () => {
       providers: [BB],
       exports: [BB],
     })
-    class B {}
+    class B implements OnModuleInit {
+      async onModuleInit() {
+        await sleep(1);
+        order.push('B');
+      }
+    }
 
     @Injectable()
     class AA implements OnModuleInit {
@@ -66,7 +74,12 @@ describe('OnModuleInit', () => {
       imports: [B],
       providers: [AA],
     })
-    class A {}
+    class A implements OnModuleInit {
+      async onModuleInit() {
+        await sleep(1);
+        order.push('A');
+      }
+    }
 
     const module = await Test.createTestingModule({
       imports: [A],
@@ -77,5 +90,7 @@ describe('OnModuleInit', () => {
 
     const instance = module.get(AA);
     expect(instance.field).to.equal('b-field_a-field');
+    expect(order[0]).to.equal('B');
+    expect(order[1]).to.equal('A');
   });
 });
```

**Comment by j** [comment]
Try adding a 3rd module.  I think it’s the right order with two but. It
deeply nested? I’ll try and make a broken test tomorrow if you don’t have
time.  Thanks for looking at it!

**Comment by micalevisk** [comment]
@j even with 3 modules. **A imports B imports C**, both with an async `onModuleInit` results in `C, B, A`, which is right :/

---

btw looks like the method `callInitHook` (`nest/packages/core/nest-application-context.ts`) is where you must look for if you want to try to fix this bug

v7:
https://github.com/nestjs/nest/blob/98eee426df5514b2194f50c9a81d8d9ac1540b55/packages/core/nest-application-context.ts#L220-L225

v8:
https://github.com/nestjs/nest/blob/8b9b0dfe7b6eef7491e4146b04d4caf1a7d25030/packages/core/nest-application-context.ts#L224-L229

**Comment by micalevisk** [comment]
@jmcdo29 I'm not that into of nest's codebase but I discoverd something that I hope help you guys.

Since the working version (v7.6.18) uses `1` as a default value for `distance`

https://github.com/nestjs/nest/blob/98eee426df5514b2194f50c9a81d8d9ac1540b55/packages/core/scanner.ts#L288

while v8.0.3 uses `0` _(and I didn't figured out why so my guess could be wrong)_
https://github.com/nestjs/nest/blob/add7eefeecd5597c5dc038daa70c3d5daed932cb/packages/core/scanner.ts#L295

thus I just made this minor regression and then this bug was gone. Note that I didn't tested anything besides the repro code that @j made.

**Comment by kamilmysliwiec** [comment]
Thanks for the reproduction repository @j!
Amazing investigation @micalevisk :tada: Fixed in 8.0.4

**Comment by j** [comment]
This is amazing to see done so quick! Thanks everyone.
